### ISSUE TITLE
Regression in mongomock 3.12.0, when updating a value to a Counter() object

### ISSUE THREAD

**Issue by boxmansf** [mongomock/mongomock#450]

**Comment by boxmansf** [comment]
Missed something, not a mongomock issue...

**Comment by boxmansf** [comment]
The issue was when a Counter() object was inserted as a document into mongodb - mongomock 3.12.0 doesn't support it

**Comment by boxmansf** [comment]
@pcorpet Actually I'm not sure - can we consider this as a regression ? As it does seem to be working in real pymongo..

**Comment by pcorpet** [comment]
Hi @boxmansf , if it's working with `pymongo`, and it worked with `mongomock` earlier versions, then yes it's a regression. Thanks for reporting;

**Comment by pcorpet** [comment]
I guess that you're writing about a `collections.Counter` object, please correct me if I'm wrong.

**Comment by boxmansf** [comment]
Yes, that’s correct

**Comment by pcorpet** [comment]
Steps to reproduce (`pymongo` needs to be installed)

```py
>>> import collections
>>> import mongomock
>>> col = mongomock.MongoClient().db.test
>>> col.insert_one({})
>>> col.update_one({}, {'$set': {'a': collections.Counter({'b': 2})}})
bson.errors.InvalidDocument: documents must have only string keys, key was ('b', 2)
```

I know where to fix it.

**Comment by pcorpet** [comment]
Sorry @boxmansf for the regression. Here's what happened:
- I've fixed a bug where OrderedDict were saved in mongomock without keeping their order
- unfortunately I did an optimization to make it easier to read the code so that it stored the actual type of the input object if it looked like a dict (so Counter got stored as counter)
- however there's a BSON check before saving that verifies that the type is compatible with BSON (which Counter is not).

I'm trying a patch where I'm keeping OrderedDict and dict, but not subclasses. I've also added a test to prevent further regressions.
